sky/skylet/job_lib.py

```diff
@@ -582,23 +580,41 @@ def update_job_status(job_ids: List[int],
                     # was killed before the job is submitted. We should set it
                     # to FAILED then. Note, if ray job indicates the job is
                     # running, we will change status to PENDING below.
-                    echo(f'INIT job {job_id} is stale, setting to FAILED')
-                    status = JobStatus.FAILED
+                    echo(
+                        f'INIT job {job_id} is stale, setting to FAILED_DRIVER')
+                    status = JobStatus.FAILED_DRIVER
 
             try:
-                # Querying status within the lock is safer than querying
-                # outside, as it avoids the race condition when job table is
-                # updated after the ray job status query.
-                # Also, getting per-job status is faster than querying all jobs,
-                # when there are significant number of finished jobs.
-                # Reference: getting 124 finished jobs takes 0.038s, while
-                # querying a single job takes 0.006s, 10 jobs takes 0.066s.
-                # TODO: if too slow, directly query against redis.
-                ray_job_status = job_client.get_job_status(ray_job_id)
-                status = _RAY_TO_JOB_STATUS_MAP[ray_job_status.value]
-            except RuntimeError:
-                # Job not found.
-                pass
+                # job_pid is 0 if the job is not submitted yet.
+                # job_pid is -1 if the job is submitted with SkyPilot older than
+                # #4318, using ray job submit. We skip the checking for those
+                # jobs.
+                if job_pid > 0:
+                    if _is_job_driver_process_running(job_pid, job_id):
+                        status = JobStatus.PENDING
+                    else:
+                        # By default, if the job driver process does not exist,
+                        # the actual SkyPilot job is one of the following:
+                        # 1. Still pending to be submitted.
+                        # 2. Submitted and finished.
+                        # 3. Driver failed without correctly setting the job
+                        #    status in the job table.
+                        # Although we set the status to FAILED_DRIVER, it can be
+                        # overridden to PENDING if the job is not submitted, or
+                        # any other terminal status if the job driver process
+                        # finished correctly.
+                        status = JobStatus.FAILED_DRIVER
```

Comment:
To help debugging in the future, would it be helpful to add debug logging at all state transition that lead to `FAILED_DRIVER` state?